subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java

```diff
@@ -121,7 +124,8 @@ public void call(SourceUnit source, GeneratorContext context, ClassNode classNod
         }
 
         private void inspectClassNode(SourceUnit sourceUnit, ClassNode classNode) {
-            sourceClassesMapping.put(new File(sourceUnit.getSource().getURI().getPath()), classNode.getName());
+            File relativePath = compilationSourceDirs.relativize(new File(sourceUnit.getSource().getURI().getPath())).orElseThrow(IllegalStateException::new);
+            sourceClassesMapping.put(relativePath, classNode.getName());
```

Comment:
Let's use the relative path as `String` and not as `File` as key to the map.

Reply:
Done.